plugins/worker/lib/BlockedBy.js

```diff
@@ -154,176 +165,199 @@ async function checkMultipleBuildsInSameEvent(buildConfig, blockingBuildIds, bui
+    // Register the curent job as running by setting key
+    await this.queueObject.connection.redis.set(runningKey, buildId);
+    // Set lastRunningKey
+    await this.queueObject.connection.redis.set(lastRunningKey, buildId);
 
-            // If any blocking job is running, then re-enqueue
-            if (blockingBuildIds.length > 0) {
-                // if build is from same event then don't re-enqueue
-                const isSameBuild = await checkMultipleBuildsInSameEvent(
-                    buildConfig,
-                    blockingBuildIds,
-                    buildId,
-                    this.queueObject.connection.redis
-                );
+    // Set expire time to take care of the case where
+    // afterPerform failed to call and blocked jobs will be stuck forever
+    await this.queueObject.connection.redis.expire(runningKey, this.blockTimeout(timeout) * 60);
+    await this.queueObject.connection.redis.expire(lastRunningKey, this.blockTimeout(timeout) * 60);
 
-                if (isSameBuild) {
-                    return false;
-                }
-                if (enforceBlockedBySelf && collapse) {
-                    await collapseBuilds.call(this, {
-                        waitingKey,
-                        buildId,
-                        blockingBuildIds
-                    });
-                } else {
-                    await this.reEnqueue(waitingKey, buildId, blockingBuildIds);
-                }
+    // Proceed
+    return true;
+}
 
-                return false;
-            }
-        }
+class BlockedBy extends NodeResque.Plugin {
+    /**
+     * Construct a new BlockedBy plugin
+     * @method constructor
+     */
+    constructor(worker, func, queue, job, args, options) {
+        super(worker, func, queue, job, args, options);
 
-        if (enforceBlockedBySelf) {
-            // only check this if feature is on
-            const blocked = await blockedBySelf.call(this, {
-                // pass in this context
-                waitingKey,
-                buildId,
-                runningBuildId,
-                collapse
-            });
+        this.name = 'BlockedBy';
+    }
 
-            if (blocked) {
-                return false;
-            } // if blocked then cannot proceed
-        } else {
-            // clean up waitingKey
-            await this.queueObject.connection.redis.del(waitingKey);
-        }
+    /**
+     * Returns true to proceed
+     * @method beforePerform
+     * @return {Promise}
+     */
+    async beforePerform() {
+        const { jobId, buildId } = this.args[0];
+        let shouldProceed = false;
 
-        // Register the curent job as running by setting key
-        await this.queueObject.connection.redis.set(runningKey, buildId);
-        // Set lastRunningKey
-        await this.queueObject.connection.redis.set(lastRunningKey, buildId);
+        try {
+            const lock = await redlock.lock(`jobId_${jobId}`, REDIS_LOCK_TTL);
```

Comment:
Do we need to explicitly check if a lock can be acquired? What happens when we cannot, does this process return exeception

Reply:
it will retry 5 times base on config from line 5, throws exception when it redis fails